torch/nn/modules/transformer.py

```diff
@@ -15,6 +15,21 @@
 
 __all__ = ['Transformer', 'TransformerEncoder', 'TransformerDecoder', 'TransformerEncoderLayer', 'TransformerDecoderLayer']
 
+def _generate_square_subsequent_mask(
+        sz: int,
+        device: torch.device = torch.device('cpu'),
+        dtype: torch.dtype = torch.get_default_dtype(),
+) -> Tensor:
+    r"""Generate a square causal mask for the sequence. The masked positions are filled with float('-inf').
+        Unmasked positions are filled with float(0.0).
+    """
+    return torch.triu(
+        torch.full((sz, sz), float('-inf'), dtype=dtype, device=device),
+        diagonal=1,
+    )
```

Comment:
Hey, in an unrelated project I just noticed that `torch.triu` cannot be applied to `bfloat16` CUDA tensors on at least PyTorch 2.0. I don't know if that changed in the meantime.

For efficiency we could do an if-query on the dtype and
- if it is `bfloat16`, move the tensor to `bfloat16` after `torch.triu` has been applied using `torch.Tensor.to`,
- otherwise (the tensor is not `bfloat16`) create it with the correct dtype immediately in `torch.full` as it is being done now.

And I guess we need to add a test that uses other dtypes to catch these things earlier.

Reply:
BF16 is a relatively new data and it is missing various support functions, esp on CPU (I recall when I tried a few years ago FP16 was also missing support on CPU for many operators.)

Is this on CPU or GPU or both?  The right course of action may be to file an issue documenting this and showing your use cases, so it might be prioritized, rather than tiptoeing around it and putting a lot of effort (and if spaghetti)! To special case some data types.

For CPU, I believe I saw messages from Intel about adding support - if you done the appropriate name in our issues you might tag that contributor. If you attach a repro, that would might help that issue.

Reply:
While we're on the definition of `_generate_square-ubseqent_mask`, I do wonder whether we should have a global function `F.generate_(square?)_subsequent_mask`n as a global.  The definition of creating one isn't majorly involved, but it crops up all over the place,  and having a single definition might make it easier for torch.compile() to optimize the code generation, as well as recognize comparisons of the mask against itself.

Of course, this might also done in overrides.py at the IR-level, by recognizing patterns and canonicalizing them?

Reply:
Hey! I checked and the issue is only on GPU; on CPU it works. :)
I'll try to search for an issue and upvote it if I find something! If I don't, I'll follow your suggestion and open one.

I don't know enough about how this would affect TorchScript/Dynamo; shouldn't the definition in `Transformer` already be basically a singleton? I guess the major question is whether it's worth breaking backward-compatibility for the change (if you implied to remove `Transformer.generate_square_subsequent_mask`).

IMO the current way is fine, but again – I don't know how/if it affects the compiler backends. :)

Reply:
Actually regarding bfloat16 and `torch.triu`, it seems to already be fixed upstream! https://github.com/pytorch/pytorch/pull/101414
So it's perfect that everything stayed as it is. :)